tests/kernel/irq_lock/src/main.c

```diff
@@ -0,0 +1,134 @@
+#include <ztest.h>
+
+#define  STACK_SIZE 512
+
+static struct k_sem sema;
+
+static K_THREAD_STACK_DEFINE(test_thread_stack, STACK_SIZE);
+static struct k_thread test_thread_d;
+
+#define NUM_SECONDS(x) ((x) * 1000)
+#define ISR0_OFFSET 10
+#define ISR2_OFFSET 11
+
+#define IRQ_LINE(offset) (CONFIG_NUM_IRQS - ((offset) + 1))
+#define NUM_SECONDS(x) ((x) * 1000)
+
+volatile u32_t val;
+volatile u32_t handler_executed;
+u32_t VAL = 0xDEAD;
+int key;
+
+
+#if defined(CONFIG_ARM)
+#include <arch/arm/cortex_m/cmsis.h>
+void trigger_irq(int irq)
+{
+	printk("Triggering irq %d\n", irq);
+#if defined(CONFIG_SOC_TI_LM3S6965_QEMU) || defined(CONFIG_CPU_CORTEX_M0) \
+	|| defined(CONFIG_CPU_CORTEX_M0PLUS)
+	NVIC_SetPendingIRQ(irq);
```

Comment:
AFAIK this should work for a Cortex-M4.

Reply:
For Cortex-M4 and others cortex series except Cortex_M0 and Cortex_M0PLUS we are using Software Trigger Interrupt Register(STIR) to trigger the interrupt. After writing the irq number in STIR register, we are triggering the interrupt.
NVIC->STIR = irq;